### ISSUE TITLE
Different behavior when attempting to fetch kind binary

### ISSUE THREAD

**Issue by shashank-g172** [kubernetes-sigs/kind#871]
**What happened**: I'm trying to create a docker image that can execute kind commands and create clusters on the host machine. I'm trying to do this in an automated way (without having to worry about having go on your path, or fetching the kind binary every time) via a maven project. Here's my expected sequence of events:

go get sigs.k8s.io/kind
(from the base directory) cd /src/sigs/k8s.io
go build
cd ../bin
get the kind executable which is just built and move it to a docker container so I can then not have to worry about fetching the kind binary every time and the docker image itself can be self-maintained.

The way I'm doing all of this is via the maven-exec plugin. When I simply run go get sigs.k8s.io/kind, I can see the folders show up where I specify the GOPATH to be at. When I try to do the same thing via the maven exec plugin, I see the following error:

package sigs.k8s.io/kustomize/pkg/commands/build: cannot find package "sigs.k8s.io/kustomize/pkg/commands/build" in any of:
        /home/{uName}/go/src/sigs.k8s.io/kustomize/pkg/commands/build (from $GOROOT)
        /home/sandbox/{uName}/workspace/funStuff/get-kind-with-exec/src/sigs.k8s.io/kustomize/pkg/commands/build (from $GOPATH)
17353 [INFO] ------------------------------------------------------------------------
17354 [INFO] BUILD FAILURE
17356 [INFO] ------------------------------------------------------------------------
17357 [INFO] Total time:  16.659 s
17357 [INFO] Finished at: 2019-09-26T08:51:03-04:00
17358 [INFO] ------------------------------------------------------------------------
17359 [ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.3.2:exec (get-kind) on project get-kind-with-exec: Command execution failed.: Process exited with an error: 1 (Exit value: 1) -> [Help 1]

This is the exec-plugin setup:

 <properties>
        <currentVersion>${project.version}</currentVersion>
        <GOROOT>/home/{Uname}/go</GOROOT>
        <GOEXEC>${GOROOT}/bin/go</GOEXEC>
        <GOBIN>${project.build.directory}</GOBIN>
</properties>

<plugins>
            <plugin>
                <groupId>org.codehaus.mojo</groupId>
                <artifactId>exec-maven-plugin</artifactId>
                <version>1.3.2</version>
                <configuration>
                    <environmentVariables>
                        <GOPATH>${basedir}</GOPATH>
                        <GOROOT>${GOROOT}</GOROOT>
                        <GOBIN>${GOBIN}</GOBIN>
                    </environmentVariables>
                </configuration>
                <executions>
                    <execution>
                        <id>get-kind</id>
                        <phase>generate-resources</phase>
                        <goals>
                            <goal>exec</goal>
                        </goals>
                        <configuration>
                            <executable>${GOEXEC}</executable>
                            <arguments>
                                <argument>get</argument>
                                <argument>sigs.k8s.io/kind</argument>                                
                            </arguments>
                        </configuration>
                    </execution>                   
                </executions>
            </plugin>

and the variables GOROOT and GOEXEC point to the same location from where I've normally set my go executables when I run go from the command line. 

**What you expected to happen**: The exec-plugin behavior to be the equivalent of "go get sigs.k8s.io/kind which results in the commands/build executable under kustomize being found.

**How to reproduce it (as minimally and precisely as possible)**: Create a maven project, point the GOEXEC and GOROOT variables and copy the maven-exec plugin execution from above to reproduce the error.

**Anything else we need to know?**: This is the first execution of the 4 steps that I've mentioned above. The second execution of the exec plugin would be to actually build the binary.  If there's a quicker and easier way to automate using Kind without downloading the binary or having go on the path - I'd be open to listening to that as well. 

**Environment:** Linux

- kind version: (use `kind version`): 0.15.1
- Kubernetes version: (use `kubectl version`):
- Docker version: (use `docker info`): 
- OS (e.g. from `/etc/os-release`): Linux

**Comment by BenTheElder** [comment]
both places we have install documentation cover this:
https://github.com/kubernetes-sigs/kind#installation-and-usage
https://kind.sigs.k8s.io/docs/user/quick-start/#installation

kind requires an official supported version of go https://golang.org/doc/devel/release.html#policy
kind is developed with go modules https://github.com/golang/go/wiki/Modules

go modules must be enabled when getting kind, because some of our dependencies use modules

this part:
> package sigs.k8s.io/kustomize/pkg/commands/build: cannot find package "sigs.k8s.io/kustomize/pkg/commands/build" in any of:
/home/{uName}/go/src/sigs.k8s.io/kustomize/pkg/commands/build (from $GOROOT)
/home/sandbox/{uName}/workspace/funStuff/get-kind-with-exec/src/sigs.k8s.io/kustomize/pkg/commands/build (from $GOPATH)

is due to not enabling go modules (which are automatically on in 1.13)

Alternatively, you can use the makefile to build in a docker container automatically with the preferred version of go used by the project. Details are in https://kind.sigs.k8s.io/docs/user/quick-start/#installation

For CI however I recommend just fetching a binary release as this is the quickest option versus repeatedly compiling. We also have instructions for this in the install instructions.

See also: https://kind.sigs.k8s.io/docs/user/resources/#using-kind-in-ci

**Comment by BenTheElder** [comment]
>If there's a quicker and easier way to automate using Kind without downloading the binary or having go on the path - I'd be open to listening to that as well.

`make build` (also mentioned in the docs) .

I'd still recommend downloading the binary though, the builds are reproducible and statically linked so you can verify the binary before starting to use it in CI.

I think this is pretty well covered between the readme and the docs, so I'm going to close this now, but if we've missed something please re-open or file a new issue.

**Comment by shashank-g172** [comment]
Thank you very much @BenTheElder. Updating to go 1.13 worked.

**Comment by shashank-g172** [comment]
I think I spoke too soon. It worked the first couple of times I ran this experiment, but it started failing again even with the 1.13.1 version. I'm on a Debian 9 machine. Here's my output:

1184 [INFO] --- exec-maven-plugin:1.3.2:exec (get-version-number) @ get-kind-with-exec ---
go version go1.13.1 linux/amd64
2742 [INFO]
2743 [INFO] --- exec-maven-plugin:1.3.2:exec (get-kind) @ get-kind-with-exec ---
package sigs.k8s.io/kustomize/pkg/commands/build: cannot find package "sigs.k8s.io/kustomize/pkg/commands/build" in any of:
        /home/{userName}/go/go/src/sigs.k8s.io/kustomize/pkg/commands/build (from $GOROOT)
        /devel/sandbox/{userName}/workspace/funStuff/get-kind-with-exec/src/sigs.k8s.io/kustomize/pkg/commands/build (from $GOPATH)

I realize this probably isn't much to go off of, but this is as basic a POM file as I can construct to see if anybody can reproduce the issue:

<plugins>
            <plugin>
                <groupId>org.codehaus.mojo</groupId>
                <artifactId>exec-maven-plugin</artifactId>
                <version>1.3.2</version>
                <configuration>
                    <environmentVariables>
                        <GOPATH>${basedir}</GOPATH>
                        <GOROOT>home/{userName}/go</GOROOT>
                        <GOBIN>${project.build.directory}</GOBIN>
                    </environmentVariables>
                </configuration>
                <executions>
                	<execution>
                		<id>get-version-number</id>
                		<phase>generate-resources</phase>
                		<goals>
                			<goal>exec</goal>
                		</goals>
                		<configuration>
                			<executable>${GOEXEC}</executable>
                			<arguments>
                				<argument>version</argument>
                			</arguments>
                		</configuration>
                	</execution>
                    <execution>
                        <id>get-kind</id>
                        <phase>generate-resources</phase>
                        <goals>
                            <goal>exec</goal>
                        </goals>
                        <configuration>
                            <executable>${GOEXEC}</executable>
                            <arguments>
                                <argument>get</argument>
                                <argument>sigs.k8s.io/kind</argument>                               
                            </arguments>
                        </configuration>
                    </execution>                   
                </executions>
            </plugin>

It worked the first 2 times, and then I cleared out my target directory and tried again and started failing the same way again.
